File: sorting-searching/sorting-algorithms/tree-sort/tree-sort.cpp
#include <iostream>
#include <vector>
using namespace std;

class Node {
public:
    int value;
    Node* left;
    Node* right;

    Node(int val) {
        value = val;
        left = nullptr;
        right = nullptr;
    }
};

Node* insert(Node* root, int value) {
    if (root == nullptr) {
        return new Node(value);
    }

    if (value < root->value) {
        root->left = insert(root->left, value);
    } else {
        root->right = insert(root->right, value);
    }

    return root;
}

void inorder(Node* root, vector<int>& result) {
    if (root == nullptr) return;

    inorder(root->left, result);
    result.push_back(root->value);
    inorder(root->right, result);
}

vector<int> treeSort(vector<int>& arr) {
    Node* root = nullptr;

    for (int value : arr) {
        root = insert(root, value);
    }

    vector<int> result;
    inorder(root, result);

    return result;
}

int main() {
    vector<int> arr = {5, 3, 7, 2, 4, 6, 8};

    vector<int> sorted = treeSort(arr);

    for (int value : sorted) {
        cout << value << " ";
    }

    return 0;
}

File: sorting-searching/sorting-algorithms/tree-sort/tree-sort.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def insert(root, value):
    if root is None:
        return Node(value)

    if value < root.value:
        root.left = insert(root.left, value)
    else:
        root.right = insert(root.right, value)

    return root


def inorder(root, result):
    if root:
        inorder(root.left, result)
        result.append(root.value)
        inorder(root.right, result)


def tree_sort(arr):
    root = None

    for value in arr:
        root = insert(root, value)

    result = []
    inorder(root, result)
    return result


if __name__ == "__main__":
    arr = [5, 3, 7, 2, 4, 6, 8]
    print("Original:", arr)
    print("Sorted:", tree_sort(arr))
